File: javascript/problems/two_pointers/validPalindrome.js
/**
 * Valid Palindrome
 *
 * Return true when a string reads the same forward and backward after we
 * ignore punctuation, spaces, and uppercase/lowercase differences.
 *
 * Example: "A man, a plan, a canal: Panama" is true.
 *
 * n = number of characters in s.
 *
 * How to test:
 * Test file: tests/javascript/validPalindrome.test.js
 * Run: node --test tests/javascript/validPalindrome.test.js
 */

function isAlphanumeric(char) {
  const code = char.charCodeAt(0);

  return (
    (code >= "A".charCodeAt(0) && code <= "Z".charCodeAt(0)) ||
    (code >= "a".charCodeAt(0) && code <= "z".charCodeAt(0)) ||
    (code >= "0".charCodeAt(0) && code <= "9".charCodeAt(0))
  );
}

/**
 * Simple solution: make a cleaned lowercase string, reverse it, and compare.
 *
 * Time: O(n), because we read the string a few times.
 * Space: O(n), because cleaned and reversed strings are created.
 */
function isPalindromeBruteForce(s) {
  const cleaned = s
    .toLowerCase()
    .split("")
    .filter(isAlphanumeric)
    .join("");

  const reversed = cleaned.split("").reverse().join("");
  return cleaned === reversed;
}

/**
 * Optimized: compare from both ends without building another string.
 *
 * left moves from the beginning; right moves from the end. Both skip
 * punctuation and spaces. A mismatch means the string is not a palindrome.
 *
 * Time: O(n): each pointer only moves inward.
 * Space: O(1): only two indexes are stored.
 */
function isPalindrome(s) {
  let left = 0;
  let right = s.length - 1;

  while (left < right) {
    while (left < right && !isAlphanumeric(s[left])) {
      left += 1;
    }

    while (left < right && !isAlphanumeric(s[right])) {
      right -= 1;
    }

    if (s[left].toLowerCase() !== s[right].toLowerCase()) {
      return false;
    }

    left += 1;
    right -= 1;
  }

  return true;
}

module.exports = { isPalindrome, isPalindromeBruteForce };

File: python/problems/two_pointers/valid_palindrome.py
"""Valid Palindrome.

Return True when a string reads the same forward and backward after ignoring
punctuation, spaces, and uppercase/lowercase differences.

Example: "A man, a plan, a canal: Panama" is True.

n = number of characters in s.

How to test:
Test file: tests/python/test_valid_palindrome.py
Run: PYTHONDONTWRITEBYTECODE=1 python3 -m unittest -v tests/python/test_valid_palindrome.py
"""


def is_alphanumeric(char: str) -> bool:
    """Return True only for English letters A-Z/a-z or digits 0-9."""
    return (
        ord("A") <= ord(char) <= ord("Z")
        or ord("a") <= ord(char) <= ord("z")
        or ord("0") <= ord(char) <= ord("9")
    )


def is_palindrome_brute_force(s: str) -> bool:
    """Clean the string, reverse it, then compare the two strings.

    Time: O(n), because we read the string a few times.
    Space: O(n), because cleaned and reversed strings are created.
    """
    cleaned_chars = []

    for char in s:
        # Python's built-in isalnum() is true for letters and digits.
        if char.isalnum():
            cleaned_chars.append(char.lower())

    cleaned = "".join(cleaned_chars)
    return cleaned == cleaned[::-1]


class Solution:
    def isPalindrome(self, s: str) -> bool:
        """Compare characters from both ends without making a cleaned string.

        left and right skip punctuation/spaces, then compare lowercase letters.

        Time: O(n): each pointer only moves inward.
        Space: O(1): only two indexes are stored.
        """
        left, right = 0, len(s) - 1

        while left < right:
            # Move left past spaces and punctuation.
            while left < right and not is_alphanumeric(s[left]):
                left += 1

            # Move right past spaces and punctuation.
            while left < right and not is_alphanumeric(s[right]):
                right -= 1

            # Compare without caring about uppercase/lowercase differences.
            if s[left].lower() != s[right].lower():
                return False

            left += 1
            right -= 1

        return True
